design/host-profiles.md

```diff
@@ -0,0 +1,2008 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# host-profiles
+
+## Status
+
+One of: provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [host-profiles](#host-profiles)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories [optional]](#user-stories-optional)
+            * [Controlling Types of Hosts Added to the Cluster](#controlling-types-of-hosts-added-to-the-cluster)
+            * [Testing On New Host Types](#testing-on-new-host-types)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+         * [Example Ironic Inventory Results](#example-ironic-inventory-results)
+            * [Virtual Machine](#virtual-machine)
+            * [Bare Metal](#bare-metal)
+
+<!-- Added by: stack, at: 2019-02-15T16:58-05:00 -->
+
+<!--te-->
+
+## Summary
+
+Host profiles allow us to match the hardware characteristics of a host
+to provide a label similar to an instance flavor in a traditional
+cloud environment. Host profiles are configurable through the
+Kubernetes API, allowing cluster administrators to expand the known
+hardware platforms and control how they are used in the cluster.
+
+## Motivation
+
+We want to limit the provisioning operations to known hardware
+platforms, while making the set of known platforms easy to expand over
+time. We do not want to hard-code the definitions, because that makes
+it more difficult to configure development or test environments that
+may not match what an expected production environment looks like.
+
+### Goals
+
+- Provide a place for us to store known hardware profiles.
+- Provide a way to match known profiles to actual hardware.
```

Comment:
Do we feel that matching is analogous to validation? In other words, when I want to ensure a host  has 5 disks (4 is not acceptable) to allow it to be provisionable is this performed by way of this host lacking the target profile because it doesn't fit the criteria or would we need to account for this separately?  While I could see how the binary form of achieving the profile or not could be validation, I would worry it doesn't quite get me what I want in terms of validation which is clear and concise reporting to operators; namely, that I expected this host to have A, but I found only B through introspection.

Reply:
Great question. For my own use cases, I've been considering the profile match as boolean, where a failure to match any known profile means the system doesn't know what the hardware is well enough to be able to use it properly. Exactly what *that* means depends on how detailed the hardware profiles are. I have not been considering exact number of disks is a useful matching aspect, though a minimum number makes sense.

If we have no profile match, I could see reporting why each failed to match. OTOH, if there are a lot of profiles, that would be very noisy.

We could do something like say if the user provides a profile name, then we report a more detailed error about why that profile does not match. That's not exactly an ideal UX, though, since it requires additional work from the user to get that data.

I'm definitely open to suggestions on how to make that validation work in a more friendly way.

Reply:
Oh I see the profile matching Boolean discussion is happening here. I agree that putting in the diffs between the host and each profile is verbose. Could possibly provide the top 3 unique profiles ranked by fewest differences in configuration to most. I could see this working well as a way to aid operators in diagnosing why a host doesn't match a previously matched profile, only if the assumption that nothing major about the host would change in terms of what of the host is represented in the profile, outside of a reprovisioning. So, drift

Reply:
This is a tough question. I'm starting to think that the original intent behind matching profiles automatically and the idea of validation may require separate solutions. Maybe a second controller that looks for hosts with the "unknown" profile and does something to annotate them about why no profiles match? We may be able to reuse some of the comparison logic, but I expect the baremetal-operator logic to be quite a bit simpler if we're only looking for exact matches.